File: JAVA/Sorting Methods/MergeSort.java
import java.util.Scanner;

public class MergeSort {
    // Function that sorts the array using merge sort
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            // Recursively sort the first and second halves
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            // Merge the sorted halves
            merge(arr, left, mid, right);
        }
    }

    // Function to merge two sorted subarrays
    public static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        // Create temporary arrays
        int[] L = new int[n1];
        int[] R = new int[n2];

        // Copy data to temporary arrays
        for (int i = 0; i < n1; i++) {
            L[i] = arr[left + i];
        }
        for (int j = 0; j < n2; j++) {
            R[j] = arr[mid + 1 + j];
        }

        // Merge the temporary arrays back into the original array
        int i = 0, j = 0;
        int k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        // Copy any remaining elements of L[]
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        // Copy any remaining elements of R[]
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    public static void main(String[] args) {
        // Create a scanner object for user input
        Scanner scanner = new Scanner(System.in);

        // Ask the user for the size of the array
        System.out.print("Enter the number of elements in the array: ");
        int n = scanner.nextInt();

        // Create an array of size n
        int[] arr = new int[n];

        // Get user input for the elements of the array
        System.out.println("Enter the elements of the array:");
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextInt();
        }

        // Call mergeSort function to sort the array
        mergeSort(arr, 0, arr.length - 1);

        // Print the sorted array
        System.out.println("Sorted array: ");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }

        scanner.close();  // Close the scanner to avoid resource leak
    }
}

File: C/Sorting_Algorithms_in_C/MergeSort.c
#include<stdio.h>

void mergeSort(int a[], int low, int high);
void merge(int a[], int low, int mid, int high);

void mergeSort(int a[], int low, int high){
    int mid;
    if(low < high){
        mid = (low + high) / 2;
        mergeSort(a, low, mid);
        mergeSort(a, mid + 1, high);
        merge(a, low, mid, high);
    }
}

void merge(int a[], int low, int mid, int high){
    int i = low;
    int j = mid + 1;
    int k = low;
    int b[50];  

    // Merging the two parts into temp array `b[]`
    while(i <= mid && j <= high){
        if(a[i] < a[j]){
            b[k] = a[i];
            i++;
        }
        else{
            b[k] = a[j];
            j++;
        }
        k++;
    }

    // Copy remaining elements from the first half, if any
    while(i <= mid){
        b[k] = a[i];
        i++;
        k++;
    }

    // Copy remaining elements from the second half, if any
    while(j <= high){
        b[k] = a[j];
        j++;
        k++;
    }

    // Copy the sorted elements back into original array `a[]`
    for(i = low; i <= high; i++){
        a[i] = b[i];
    }
}

int main(){
    int a[] = {8, 3, 5, 7, 2, 1, 10, 4};
    int n = sizeof(a) / sizeof(a[0]);

    // Print the original array
    printf("Original array: \n");
    for(int i = 0; i < n; i++){
        printf("%d  ", a[i]);
    }
    printf("\n\n");

    // Apply merge sort
    mergeSort(a, 0, n - 1);

    // Print the sorted array
    printf("Sorted array: \n");
    for(int i = 0; i < n; i++){
        printf("%d  ", a[i]);
    }
    
    return 0;
}
